### DISCUSSION TITLE
How can I create/modify a Widget using C#?

### DISCUSSION THREAD

**Discussion by jackliDev** [OrchardCMS/OrchardCore#15856] [Q&A]
The examples in the [docs ](https://docs.orchardcore.net/en/dev/docs/reference/modules/Widgets/)(https://docs.orchardcore.net/en/dev/docs/reference/modules/Widgets/) use Liquid.

I haven't found any Orchard Core examples using C# for Widgets but there is a project OrchardCore.Widgets in the OrchardCore repo that contains C# code.

Is there a way to create/modify Widgets in Orchard Core using C#?

**Comment by deanmarcussen** [comment]
The docs you linked to have examples of both Razor and Liquid (you just need to click on the Razor tab) templates for widgets

I assume it is the views you want to modify?

If you want to template with Razor rather than Liquid just create a `.cshtml` file as per the above docs

**Comment by jackliDev** [comment]
Hi @deanmarcussen , thanks for the response. I'm trying to find out if there is a way to create/modify widgets using C# specifically?

**Comment by deanmarcussen** [comment]
Normally widgets are contained inside another content item, like a FlowPart, or a WidgetList, or the Layers, so it would be unusual to programatically try and create one standalone (as you couldn't do anything with it).

You might be asking if you can create a FlowPart programatically with some widgets inside it?

Or you might be asking if you can create a Widget content definition programatically?

**Comment by jackliDev** [comment]
A bit more background info:
We're trying to create custom widgets for a variety of purposes. For example, one widget we may want to create could be a calendar and we want to this widget to be able to pull configuration data from a client's db to determine which start day of the week to show. 

Let's say one client's db has configuration info saying to show Monday as the start of week. Another client could have Sunday as the start of week. We want our custom widget to render accordingly.

Is it possible to do this or to extend the current Widget logic for this purpose?

Also, I've been trying to hit the code in the OrchardCore.Widgets project. However, whether I create a widget or render it, nothing I do hit my breakpoints. What user actions will hit the code in this project ?

**Comment by deanmarcussen** [comment]
Ok, so yes you could use a `Widget` for that - by applying an alternate view to the Widget template, like for example `Widget-Calendar.cshtml`

I think what you might be misunderstanding is how Widgets are applied / used.

Just creating one as a normal content item will not do anything.

You need to add them to a Content Item which has either a `FlowPart`, `WidgetListPart` or one of the `Layers`.

Then if you have not made an alternate like above, the `Widget.cshtml` template in the `OrchardCore.Widgets` will be used.

Recommend you take a look at the BlogTheme, and try making a `Page` which has a `FlowPart` attached, allowing you to add widgets to the `FlowPart`, to get a better understanding of how they are used.   

Does that help?

**Comment by sebastienros** [comment]
What @deanmarcussen means is that you need to create a new "Content Type" and then set `Widget` as the stereotype. Then you can create the alternates he mentions.

**Comment by webmedia1012** [comment]
I see this item is not yet closed, that's a good thing I wanted to clarify what @deanmarcussen  mentioned, I'm using a headless approach vs the blogtheme, and I understand how this works with that recipe, but what if I'm using the helper (IOrcharHelper)? what method invokes the widget ? that's available to me?

**Comment by sebastienros** [comment]
Not closed, because still no code sample in the doc.

Just some clarification, is it about creating a new Widget Type dynamically, or creating a new widget instance for a known type?
